lib/SILOptimizer/Mandatory/TFPartition.cpp

```diff
@@ -997,27 +1003,28 @@ void TFFunctionPartition::markBlock(SILBasicBlock *BB) {
         }
       }
 
-      // Check to see if pred is already visited or if its terminator is already
-      // marked then we don't need to reprocess it.
-      if (visited.count(pred) ||
-          markedInstructions.count(pred->getTerminator()))
-        continue;
-
-      // If the predecessor has a single tensor-related successor (us) then we
-      // aren't control-dependent on it.  It may be control-dependent on
-      // something though.
       if (numTensorSuccs == 1) {
-        assert(tensorCodeBlocks.postDominates(thisBB, pred));
-        assert(tensorCodeBlocks.postDominates(BB, pred));
-        worklist.push_back(pred);
+        // Some sanity checks.
+        assert(tensorCodeBlocks.properlyPostDominates(thisBB, pred));
+        assert(tensorCodeBlocks.properlyPostDominates(BB, pred));
+      }
+      // We aren't control-dependent on `pred`, if `pred` is just another block
+      // post-dominated by `BB`. This includes the case where `pred` has a
+      // single tensor-related successor (us).
+      // In those cases, continue walking `pred` if we haven't processed it,
+      // since it may be control-dependent on something.
+      if (tensorCodeBlocks.properlyPostDominates(BB, pred)) {
+        if (!addedToList.count(pred)) {
```

Comment:
doing count followed by insert does two hash table lookups.  It is better to just do "insert" and check the bool result to see if it was already there.

Reply:
Good point. changed.